extensions/cl_khr_command_buffer_internal_buffer.asciidoc

```diff
@@ -0,0 +1,256 @@
+}
+// -->
+void a_command_buffer() {
+  cl_mem buf0 = ..., buf1 = ...;
+  kernelA(in, buf0);
+  kernelB(buf0, buf1);
+  kernelC(buf1, buf0)
+  kernelD(buf0, out);
+}
+----
+* fuse kernels together as intermediate results do not need to be
+  preserved. In C analogy:
++
+[source,c]
+----
+cl_mem in, w, out;
+void a_command_buffer() {
+  cl_mem buf0 = ...;
+  convolutionKernel(in, w, buf0);
+  reluKernel(buf0, out);
+}
+// -->
+void a_command_buffer() {
+  convolutionPlusReluKernel(in, w, out);
+}
+----
+
+== New API Functions
+
+None.
+
+== New API Types
+
+None.
+
+== New API Enums
+
+[source,c]
+----
+CL_MEM_COMMAND_BUFFER_INTERNAL_KHR 0x????
+----
+
+== Modifications to the OpenCL API Specification
+
+(Modify Section 5.2.1, *Creating Buffer Objects*) ::
++
+--
+
+(Add the following to the table of buffer creation properties) ::
++
+--
+[cols="2,1,2",stripes=odd,options="header"]
+|===
+| Propery | Property Value | Description
+
+| `CL_MEM_COMMAND_BUFFER_INTERNAL_KHR` | `cl_khr_command_buffer` a|
+This property can be used if *cl_khr_command_buffer_internal_buffer*
+extension is supported.
+
+This property constraints the created buffer to be only accessible by
+commands recorded into the associated command buffer. Reading from or
+writing to the buffer by commands which are not part of the associated
+command-buffer is considered undefined behavior.
+
+The associated command-buffer may deallocate storage and reallocate
+the storage as needed during its execution and otherwise. Multiple
+buffers associated with the same command-buffer may share same data
+storage.
+
+// A consequence of the last sentence: CL_MEM_SIZE queries on two or
+// more buffers associated with the same command-buffer may not
+// reflect the actual storage used on during execution of the
+// command-buffer. IOW: the storage used may be lower than
+// `CL_MEM_SIZE(buf1) + CL_MEM_SIZE(buf2)`.
+
+The contents of the buffer are not guaranteed to be preserved
+after the associated command-buffer execution completes.
+
+This property is incompatible with *CL_MEM_COPY_HOST_PTR* and
+*CL_MEM_USE_HOST_PTR* memory flags and *CL_MEM_DEVICE_HANDLE_LIST_KHR*
+buffer creation property.
+
+This property implies *CL_MEM_HOST_NO_ACCESS* memory flag.
+
+The reference count of the associated command-buffer is not increased
+when the buffer is created. When the associated command-buffer is
+released the buffer becomes invalid.
+|===
+
+--
+--
+// End (Modify Section 5.2.1, *Creating Buffer Objects*)
+
+(Add to the list of error codes for *clEnqueueReadBuffer*, *clEnqueueWriteBuffer*, *clEnqueueReadBufferRect*, *clEnqueueWriteBufferRect*, *clEnqueueCopyBuffer*, *clEnqueueCopyBufferRect*, *clEnqueueFillBuffer* and *clEnqueueMapBuffer*) ::
++
+--
+* *CL_INVALID_MEM_OBJECT* if a memory object passed to this function
+  is a buffer object or references a buffer object created with
+  *CL_MEM_COMMAND_BUFFER_INTERNAL_KHR* property.
+
+// "references a buffer": E.g. sub-buffers.
```

Comment:
Do we want to allow sub-buffers to be created of these internal buffers? I wonder if there's a use case for that.

Reply:
Perhaps, we want to because OpenVX supports sub-objects of virtual objects.